### PR TITLE
[HER-53] Implement detection repo

### PR DESCRIPTION
**Pull request by ConstantinescuAndrei**
## Description
Create detection repo and register it inside IoC

## Related
[HER-53](https://nerds-interns.atlassian.net/jira/software/projects/HER/boards/1?selectedIssue=HER-53)

### CODE REVIEW

File: src/Presentation/Startup.cs [hunk 1]
```diff
@@ -24,6 +27,8 @@ public void ConfigureServices(IServiceCollection services)
             services.AddScoped<Schema>();
             services.AddScoped<Detection.Creating.IResolver, Detection.Creating.Resolver>();
             services.AddScoped<Detection.Fetching.IResolver, Detection.Fetching.Resolver>();
+            services.AddTransient<IGetDetection, GetDetection>();
+            services.AddTransient<IDetectionRepository, DetectionRepository>();
```
Comment (RaduBuciuceanu) [inline src/Presentation/Startup.cs:31]: Should we consider using the `Scoped` registrations instead?

File: src/Presentation/Detection/Fetching/Resolver.cs [hunk 2]
```diff
@@ -1,16 +1,20 @@
...
     using global::GraphQL;
 
     public class Resolver : IResolver
     {
+        private readonly IGetDetection _getDetection;
+
+        public Resolver(IGetDetection getDetection)
+        {
+            _getDetection = getDetection;
+        }
```
Comment (RaduBuciuceanu) [inline src/Presentation/Detection/Fetching/Resolver.cs:15]: Should we convert this constructor into a one liner?

File: src/Data/Detection/DetectionRepository.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+                {
+                    Id = detection.Id,
+                    Class = detection.Class,
+                    Score = detection.Score,
+                    Timestamp = detection.Timestamp,
+                };
+                detectionList.Add(detectionModel);
+            }
+
+            return detectionList;
```
Comment (RaduBuciuceanu) [inline src/Data/Detection/DetectionRepository.cs:33]: Alert, alert! C++ imperative style programming. Won't be easier to use `Linq` extensions instead?

File: src/Data/Detection/DetectionRepository.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+            var detectionList = new List<MDetection>();
+            foreach (var detection in detections)
+            {
+                var detectionModel = new MDetection
+                {
+                    Id = detection.Id,
+                    Class = detection.Class,
+                    Score = detection.Score,
+                    Timestamp = detection.Timestamp,
+                };
```
Comment (RaduBuciuceanu) [inline src/Data/Detection/DetectionRepository.cs:29]: Should we consider using a factory pattern here?

File: src/Data/Detection/DetectionRepository.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    using Business.Detection.Common.Repositories;
+
+    public class DetectionRepository : IDetectionRepository
+    {
+        private readonly Context _context;
+
+        public DetectionRepository(Context context)
+        {
+            _context = context;
+        }
```
Comment (RaduBuciuceanu) [inline src/Data/Detection/DetectionRepository.cs:14]: Should we convert this constructor into a one liner?

File: src/Data/Detection/DetectionRepository.cs (hunk 3 above)
Comment (RaduBuciuceanu) [reply src/Data/Detection/DetectionRepository.cs:33]: image

**Review by RaduBuciuceanu — CHANGES_REQUESTED** [review]

File: DetectionFactory/ConvertDetection.cs [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    {
+        [SetUp]
+        public void Setup()
+        {
+        }
+
+        [Test]
+        public void DetectionFactoryTest()
+        {
+            var actualDetection = new MDetection {
```
Comment (danidoff) [inline DetectionFactory/ConvertDetection.cs:19]: Should we also add the actualDetection and testDetection on Setup method? Let`s have a talk over this.

File: DetectionFactory/ConvertDetection.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+{
+    public class ConvertDetection
+    {
+        [SetUp]
+        public void Setup()
+        {
+        }
+
+        [Test]
+        public void DetectionFactoryTest()
```
Comment (danidoff) [inline DetectionFactory/ConvertDetection.cs:17]: Should we rename this test to look like this: "public void Method_Scenario_Expected()". Ex: "public void Post_StatusCode_IsRight()"

**Review by danidoff — CHANGES_REQUESTED** [review]

File: DetectionFactory/ConvertDetection.cs (hunk 6 above)
Comment (ConstantinescuAndrei) [reply DetectionFactory/ConvertDetection.cs:19]: Done

File: DetectionFactory/ConvertDetection.cs (hunk 7 above)
Comment (ConstantinescuAndrei) [reply DetectionFactory/ConvertDetection.cs:17]: Done

File: DetectionFactory/ConvertDetection.cs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+        {
+            var actualDetection = new MDetection {
+                Id = 1,
+                Class = "Class",
+                Score = 100
```
[the remark this answers is not in the document]
Comment (ConstantinescuAndrei) [reply DetectionFactory/ConvertDetection.cs:22]: Done

File: src/Data/Detection/DetectionRepository.cs (hunk 5 above)
Comment (ConstantinescuAndrei) [reply src/Data/Detection/DetectionRepository.cs]: Done

File: src/Data/Detection/DetectionRepository.cs (hunk 4 above)
Comment (ConstantinescuAndrei) [reply src/Data/Detection/DetectionRepository.cs]: Done

File: src/Data/Detection/DetectionRepository.cs (hunk 3 above)
Comment (ConstantinescuAndrei) [reply src/Data/Detection/DetectionRepository.cs]: Done

File: src/Presentation/Detection/Fetching/Resolver.cs (hunk 2 above)
Comment (ConstantinescuAndrei) [reply src/Presentation/Detection/Fetching/Resolver.cs:15]: Done

File: src/Presentation/Startup.cs (hunk 1 above)
Comment (ConstantinescuAndrei) [reply src/Presentation/Startup.cs]: Done

File: DetectionFactory/DetectionFactory.csproj [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+        <PackageReference Include="NUnit3TestAdapter" Version="3.16.1" />
+        <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.5.0" />
+    </ItemGroup>
+
+    <ItemGroup>
+      <ProjectReference Include="..\src\Business\Business.csproj" />
+      <ProjectReference Include="..\src\Data\Data.csproj" />
+    </ItemGroup>
+
+</Project>
```
Comment (RaduBuciuceanu) [inline DetectionFactory/DetectionFactory.csproj:23]: Should we rename this project into: `DataTests` one?

Comment (RaduBuciuceanu) [reply DetectionFactory/DetectionFactory.csproj:23]: Should we move this project under the `test` folder? Should we make sure we use solution folders also?

File: DetectionFactory/ConvertDetection.cs [hunk 10]
```diff
@@ -0,0 +1,46 @@
+using System;
+using Business.Detection.Common.Models;
+using Data;
+using Data.Detection;
+using FluentAssertions;
+using NUnit.Framework;
+
+namespace FactoryDetectionTests
+{
+    public class ConvertDetection
```
Comment (RaduBuciuceanu) [inline DetectionFactory/ConvertDetection.cs:10]: By convention the name of the test class should be composed of: `[NameOfTheClassBeingTested][Tests]`. Example:

- `FooClass` - `FooClassTests`
- `RandomGenerator` - `RandomGeneratorTests`
- `RandomFactory` - `RandomFactoryTests`

With those said, should we consider renaming this test class to match the above convention?

**Review by RaduBuciuceanu — CHANGES_REQUESTED** [review]
There is also no need for this inner `SolutionFolder` here. Should we consider removing it?

image

File: DetectionFactory/ConvertDetection.cs (hunk 10 above)
Comment (ConstantinescuAndrei) [reply DetectionFactory/ConvertDetection.cs]: Done

File: DetectionFactory/DetectionFactory.csproj (hunk 9 above)
Comment (ConstantinescuAndrei) [reply DetectionFactory/DetectionFactory.csproj]: Done

**Review by RaduBuciuceanu — CHANGES_REQUESTED** [review]
No need for those inner folders. Won't be cleaner to move the factory near to repository?
image

File: src/Data/Detection/Common/DetectionFactory/DetectionFactory.cs [hunk 11]
```diff
@@ -0,0 +1,15 @@
+namespace Data.Detection
```
Comment (RaduBuciuceanu) [inline src/Data/Detection/Common/DetectionFactory/DetectionFactory.cs:1]: Namespace does not match the folder structure. Should we make sure we don't violate this rule?

**Review by RaduBuciuceanu — CHANGES_REQUESTED** [review]

File: src/Data/Detection/Common/DetectionFactory/DetectionFactory.cs (hunk 11 above)
Comment (ConstantinescuAndrei) [reply src/Data/Detection/Common/DetectionFactory/DetectionFactory.cs]: Done

**Review by RaduBuciuceanu — APPROVED** [review]
Sexy!

**Review by danidoff — APPROVED** [review]

**Review by andreea-dumitrascu — APPROVED** [review]

**Pull request merged**